### ISSUE TITLE
Improve UX around github permissions

### ISSUE THREAD

**Issue by maxandersen** [openshiftio/openshift.io#259]
Currently we ask up front to let users grant openshift.io full permissions for github access.

Several issues raised on that:
a) there is little to no info on *why* openshift.io needs it (we need it for doing push and creation of content)
b) that it is asked as the *first* thing puts people off

Proposal is to make it so we don't request full permissions before we actually need them (at space creation or when opening/pushing in a che workspace)

**Comment by jstrachan** [comment]
things are hard enough with getting KeyCloak to sync with valid tokens from openshift + github token in a stable way as it is :). Then changing the scope dynamically to ask for more and more permissions on demand (and handling the UX workflow as we walk through the steps of asking and getting approval to change the scope) just sounds like a ton of complexity for very little actual gain leading to an overall harder to use & harder to test system. 

If folks want to use the full openshift.io platform it automates the creation of github repos & webhooks along with allowing you (and the CD pipelines) to commit to your newly created repos.

I guess if folks don't wanna create/import git repos or commit into them; we could just try make the github token totally optional. Though you can't do very much with the system if you opt out - other than raise issues.

So maybe we disable all of the Create UI (and the Forge wizard) until folks approve the github scope we really need?

**Comment by alexeykazakov** [comment]
Currently only Planner doesn't require GitHub at all (until we support github integration here), right? So, I would agree with @jstrachan that our first step could be making github linking optional on the getting started page. But then there is a question how/when we would ask users to link GitHub. By disabling the relevant UI and providing a link to the getting started page (or a new dedicated page) with explanation what (and why) to do? Or redirecting to that page automatically? Something else?

**Comment by joshuawilson** [comment]
I think we are going to need @catrobson on this one.

**Comment by jstrachan** [comment]
Maybe we just need full GitHub support to run the new app wizard or create a codebase/workspace? At least those are 2 places we could do the checks fairly easily

**Comment by gjbianco** [comment]
Even without changing the UX, it seems like some of the permissions aren't needed. For example, it doesn't seem like it needs full access to personal user data. Is read only enough?

As far as managing repos for the user, it would seem like it only needs full access to repositories.

**Comment by aslakknutsen** [comment]
Related https://github.com/openshiftio/openshift.io/issues/2118

**Comment by alexeykazakov** [comment]
> For example, it doesn't seem like it needs full access to personal user data. Is read only enough?

Yeah... "read:user" and "user:email" should be enough. We don't need write access to user profile I guess.
@pradeepto could you please confirm that you need/do not need "user" scope?

**Comment by sthaha** [comment]
I am pretty sure you would also require `read:org`.

**Comment by alexeykazakov** [comment]
yes sure. The question is if we can remove "user" scope and add "read:user" and "user:email" instead.
Currently we ask GH for `admin:repo_hook read:org repo user gist`
Could we ask for `admin:repo_hook read:org repo read:user user:email gist` instead? Will it be enough for buld?

**Comment by alexeykazakov** [comment]
@pradeepto any updates on it?

**Comment by gastaldi** [comment]
We removed the need for email in fabric8-launcher. We require at least: 
- `read:org`
- `read:user`
- `repo` (or `public_repo` is enough)
- `admin:repo_hook`

**Comment by alexeykazakov** [comment]
> - `read:org`
> - `read:user`
> - `public_repo`
> - `admin:repo_hook`

This ^ list looks better!

The list of permissions we currently require:

- `read:org`
- `user`
- `repo`
- `admin:repo_hook`
- `gist`

So, the question is if we can confirm that we can:
- drop `gist`
- replace `user` by `read:user`
- replace `repo` by `public_repo`

@gastaldi, you are confirming that the app wizard is ok with that, right?
@pradeepto, what about build?
@l0rd if there any requirement from Che at all?

**Comment by joshuawilson** [comment]
@gastaldi it's not just the new launch wizard but the old app wizard too.

**Comment by gastaldi** [comment]
@joshuawilson yes, I know that. I think it may be worth a test with the scopes that @alexey suggested, I don't think we'd need anything much different than that

**Comment by pradeepto** [comment]
@alexeykazakov @sthaha is investigating and will report back.

**Comment by sthaha** [comment]
@alexeykazakov FYI I am starting to investigate the permissions required by the build service

**Comment by sthaha** [comment]
I will start documenting the permissions required by the build service here for our ci/cd and osio

#### Jenkins: CI/CD
- github auth plugin
  - `read:org` `user:email`

#### Jenkins: OSIO

**Comment by l0rd** [comment]
@alexeykazakov in Che we use the github token to locally clone the repo and push changes to the same github repo. Hence if `read:user` is enough to push changes to the remote github repo it should be ok for Che.

**Comment by alexeykazakov** [comment]
> Hence if read:user is enough to push changes to the remote github repo it should be ok for Che.

I guess this should be enough because we still going to ask for public_repo. But can someone from the Che team confirm?

**Comment by l0rd** [comment]
@ashumilova do you think that a github token with `read:user` privilege is enough to run Che on OSIO (clone and push to remote)?

**Comment by alexeykazakov** [comment]
@l0rd @ashumilova btw, @sthaha went through tested OSIO with a GH token with reduced scopes (`admin:repo_hook read:org public_repo read:user`) yesterday. See the Che part here - https://github.com/openshiftio/openshift.io/issues/259#issuecomment-370307060
Che seemed to work well.

**Comment by alexeykazakov** [comment]
PR to reduce the asked GH permissions - https://github.com/fabric8-services/fabric8-auth/pull/394

**Comment by muruGanesan** [comment]
@alexeykazakov @sbose78, Do you need anything enhancement from UX side? e.g. while connecting with GitHub account - should we have to restrict/provide options to control the permission?
-or-
Is this handled in new permission module implementation?

**Comment by sthaha** [comment]
@alexeykazakov @sbose78 are we good to close this? 
I have removed the `build-cd` tag from this issue. Please add it back if you require more info.
